api/v1alpha1/executionhook_types.go

```diff
@@ -0,0 +1,182 @@
+
+	// This is a list of ContainerExecutionHookStatus, with each status representing
+	// information about how hook is executed in a container, including pod name,
+	// container name, ActionTimestamp, ActionSucceed, etc.
+	// +optional
+	HookStatuses []ContainerExecutionHookStatus `json:"hookStatuses,omitempty" protobuf:"bytes,1,rep,name=hookStatuses"`
+}
+
+// ContainerExecutionHookStatus represents the current state of a hook for a specific
+// container in a pod
+type ContainerExecutionHookStatus struct {
+	// This field is required
+	PodName string `json:"podName" protobuf:"bytes,1,opt,name=podName"`
+
+	// This field is required
+	ContainerName string `json:"containerName" protobuf:"bytes,2,opt,name=containerName"`
+
+	// If not set, it is nil, indicating Action has not started
+	// If set, it means Action has started at the specified time
+	// +optional
+	Timestamp *metav1.Time `json:"actionTimestamp,omitempty" protobuf:"bytes,3,opt,name=actionTimestamp"`
+
+	// Succeed is set to true when the action is executed in the container successfully.
+	// It will be set to false if the action cannot be executed successfully after
+	// ActionTimeoutSeconds passes.
+	// +optional
+	Succeed *bool `json:"actionSucceed,omitempty" protobuf:"varint,4,opt,name=actionSucceed"`
+
+	// The last error encountered when executing the action. The hook controller might
+	// update this field each time it retries the execution.
+	// +optional
+	Error *HookError `json:"error,omitempty" protobuf:"bytes,5,opt,name=error"`
+}
+
+// HookError describes the error occurred from hook execution.
+type HookError struct {
+	// Type of the error
+	// This is required
+	ErrorType ErrorType `json:"errorType" protobuf:"bytes,1,opt,name=errorType"`
+
+	// Error message
+	// +optional
+	Message *string `json:"message,omitempty" protobuf:"bytes,2,opt,name=message"`
+
+	// More detailed reason why error happens
+	// +optional
+	Reason *string `json:"reason,omitempty" protobuf:"bytes,3,opt,name=reason"`
+
+	// It indicates when the error occurred
+	// +optional
+	Timestamp *metav1.Time `json:"timestamp,omitempty" protobuf:"bytes,4,opt,name=timestamp"`
+}
+
+// ErrorType defines the type of error occurred from hook execution.
+type ErrorType string
+
+// More error types could be added, e.g., Forbidden, Unauthorized, AlreadyInProgress, etc.
+const (
+	// The execution hook times out
+	Timeout ErrorType = "Timeout"
+
+	// The execution hook fails with an error
+	Error ErrorType = "Error"
+)
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=executionhook,shortName=eh,scope=Namespaced,categories=executionhook
+// +kubebuilder:subresource:status
+
+// ExecutionHook is the Schema for the executionhook API
+type ExecutionHook struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ExecutionHookSpec   `json:"spec,omitempty"`
+	Status ExecutionHookStatus `json:"status,omitempty"`
```

Comment:
// +optional
Make Status a pointer because it is optional.  I know it is not a pointer in the original KEP, but we have discussed about this Status field a lot when promoting Snapshot to beta and determined it should be a pointer.

Add brief comments for Spec and Status